channeldb/channel.go

```diff
@@ -2718,6 +2718,12 @@ func (c *OpenChannel) CloseChannel(summary *ChannelCloseSummary,
 			return err
 		}
 
+		// Delete all the forwarding packages stored for this particular
+		// channel.
+		if err = chanState.Packager.Wipe(tx); err != nil {
```

Comment:
This should also be done in `MarkChanFullyClosed` so any channels that were fully closed after the migration will also have their forwarding packages cleared.

Reply:
`MarkChanFullyClosed` is called in two places.

First when we `ResolveContract` in `ChainArbitrator`, which is attached to `arbCfg.MarkChannelResolved`. This function is only invoked when [we reach `StateFullyResolved`](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L1069). Now, to reach `StateFullyResolved`, we must use one of the following triggers, `remoteCloseTrigger`, `localCloseTrigger`, `coopCloseTrigger` or`breachCloseTrigger`.

The formal three triggers will all invoke the method `MarkChannelClosed`, which calls `channel.CloseChannel` that will wipe the fwd packages. As for the trigger `breachCloseTrigger`, when received, we will also [close the channel via `markClosed`](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/chain_watcher.go#L1161) in `dispatchContractBreach` before we handle the case in our state machine. The call stack is somewhat deep and jumpy, as `markClosed` is again wrapped in [`processACK` in server.go](https://github.com/lightningnetwork/lnd/blob/master/server.go#L968) and passed to `ContractBreachEvent`, where this is the second place we would use `MarkChanFullyClosed`. However, before that happens, we would call [`ProcessACK`](https://github.com/lightningnetwork/lnd/blob/master/breacharbiter.go#L882) which calls `CloseChannel` that wipes the packages.

Reply:
Thanks for the through answer! One way to verify this from the PoV of an integration test would be to supplement the `PendingChannels` response w/ some information about the set of active forwarding packages. Each time we call a normal `CloseChannel` (or have a resulting closed channel and a pending channels call), we would then verify that the set of forwarding packages is 100% empty.

Reply:
Good idea. A new itest case is added!